### ISSUE TITLE
exporter supplies no metrics if redis is not reachable

### ISSUE THREAD

**Issue by lukaalba** [oliver006/redis_exporter#695]
**Describe the problem**
I am struggling with setting up proper alerts based on the redis_up metric. I've created a Redis Instance, deployed the exporter and got metrics for the redis instance. After that, I've deleted the redis instance. My assumption was, that the exporter just changes` redis_up` to 0 and continues to provide the metrics. Instead, it seems that the endpoint of the exporter becomes unavailable, because I didn't get any metrics at all. Is this intentional behavior?

The log of the exporter repeats the following bulk of messages:

```
time="2022-08-25T11:57:40Z" level=debug msg="Trying DialURL(): redis://10.143.162.75"
time="2022-08-25T11:57:55Z" level=debug msg="DialURL() failed, err: dial tcp 10.143.162.75:6379: i/o timeout"
time="2022-08-25T11:57:55Z" level=debug msg="Trying: Dial(): tcp 10.143.162.75"
time="2022-08-25T11:57:55Z" level=error msg="Couldn't connect to redis instance (10.143.162.75)"
time="2022-08-25T11:57:55Z" level=debug msg="connectToRedis( 10.143.162.75 ) err: dial tcp: address 10.143.162.75: missing port in address"
time="2022-08-25T11:57:55Z" level=debug msg="scrapeRedisHost() done"
time="2022-08-25T11:57:55Z" level=debug msg="Trying DialURL(): redis://10.143.162.75"
time="2022-08-25T11:58:10Z" level=debug msg="DialURL() failed, err: dial tcp 10.143.162.75:6379: i/o timeout"
time="2022-08-25T11:58:10Z" level=debug msg="Trying: Dial(): tcp 10.143.162.75"
time="2022-08-25T11:58:10Z" level=error msg="Couldn't connect to redis instance (10.143.162.75)"
time="2022-08-25T11:58:10Z" level=debug msg="connectToRedis( 10.143.162.75 ) err: dial tcp: address 10.143.162.75: missing port in address"
time="2022-08-25T11:58:10Z" level=debug msg="scrapeRedisHost() done"
```

**What version of redis_exporter are you running?**
Please run `redis_exporter --version` if you're not sure what version you're running.
[ ] 0.3x.x
[X] 1.x.x
v1.44.0

**Running the exporter**
redis-exporter
      --redis.addr=10.143.162.75
      --set-client-name=false
      --debug

**Screenshots**
If applicable, add screenshots to help explain your question.

**Additional context**
Exporter runs in a Kubernetes Cluster, Redis is hosted via Google Memorystore

**Comment by oliver006** [comment]
I tested locally and it works as expected, not sure why this fails.

```
time="2022-08-25T11:57:55Z" level=debug msg="Trying DialURL(): redis://10.143.162.75"
time="2022-08-25T11:58:10Z" level=debug msg="DialURL() failed, err: dial tcp 10.143.162.75:6379: i/o timeout"
```
This ^^^ shows that the exporter tries to connect to the Redis instance and then times out (15s timeout, default as per cmd line options) so this seems to work as expected.

I'm confused about your statement `continues to provide the metrics. Instead, it seems that the endpoint of the exporter becomes unavailable, because I didn't get any metrics at all.`

The exporter will stop providing most Redis instance metrics if the instance becomes unavailable though some metrics will still be available like `redis_up` and metrics about the exporter itself.

`the endpoint of the exporter becomes unavailable` - what exactly did you observe here? What happens when you use curl to access the metrics?

**Comment by lukaalba** [comment]
Thanks for the fast reply!

I have to correct myself. Opening localhost:9121 in a browser seems to work as usual, but when clicking on "metrics", it keeps loading forever. The same with curl:

```
$ curl localhost:9121/metrics -v
*   Trying 127.0.0.1:9121...
* Connected to localhost (127.0.0.1) port 9121 (#0)
> GET /metrics HTTP/1.1
> Host: localhost:9121
> User-Agent: curl/7.79.1
> Accept: */*
> 


```

**Comment by lukaalba** [comment]
I've did some additional testing with a redis instance which runs in the same kubernetes cluster as the exporter. Deployed the redis instance, started the exporter, then scaled the redis instances to 0. This test worked as expected but as soon as I removed the redis deployment completely, the behavior as described above occurred again.

It looks like the exporter handles the events "redis is not running" and "redis instance does not exist at all" differently. I assume this is intentional since it doesn't really make sense to generate metrics for a fictional redis. My test case was simply not sufficient.
